Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,201 @@
+
+Setting this environment variable will cause each of the specified
+types' `public static void Initialize()` methods to be called in
+order, synchronously, before the main assembly is loaded. The hooks
+are all called on the same managed thread (the same thread that calls
+`Main`). The environment variable will be inherited by child processes
+by default. It is up to the `StartupHook.dll`s and user code to decide
+what to do about this - `StartupHook.dll` may clear them to prevent
+this behavior globally, if desired.
+
+Specifically, hostpolicy starts up coreclr and sets up a new
+AppDomain. It then invokes a private method in
+`System.Private.CoreLib`, which will call each
+`StartupHookType.Initialize()` in turn synchronously. This gives
+`StartupHookType` a chance to set up new `AssemblyLoadContext`s, or
+register other callbacks. After all of the `Initialize()` methods
+return, control returns to hostpolicy, which then calls the main entry
+point of the app like usual.
+
+Rather than forcing all configuration to be done through a single
+predefined API, this creates a place where such configuration could be
+centralized, while still allowing user code to do its own thing if it
+so desires.
+
+The producer of `StartupHook.dll` needs to ensure that
+`StartupHook.dll` is compatible with the dependencies specified in the
+main application's deps.json, since those dependencies are put on the
+TPA list during the runtime startup, before `StartupHook.dll` is
+loaded. This means that `StartupHook.dll` needs to built against the
+same or lower version of .NET Core than the app.
+
+## Example
+
+This could be used with `AssemblyLoadContext` APIs to resolve
+dependencies not on the TPA list from a shared location, similar to
+the GAC on full framework. It could also be used to forcibly preload
+assemblies that are on the TPA list from a different location. Future
+changes to `AssemblyLoadContext` could make this easier to use by
+making the default load context or TPA list modifiable.
+
+```c#
+namespace SharedHostPolicy
+{
+    class SharedHostInitializer
+    {
+        public static void Initialize()
+        {
+            AssemblyLoadContext.Default.Resolving += SharedAssemblyResolver.LoadAssemblyFromSharedLocation;
+        }
+    }
+
+    class SharedAssemblyResolver
+    {
+        public static Assembly LoadAssemblyFromSharedLocation(AssemblyLoadContext context, AssemblyName assemblyName)
+        {
+            string sharedAssemblyPath = // find assemblyName in shared location...
+            if (sharedAssemblyPath != null)
+                return AssemblyLoadContext.Default.LoadFromAssemblyPath(sharedAssemblyPath)
+            return null;
+        }
+    }
+}
+```
+
+## Error handling details
+
+Problems with the startup hook should be fairly straightforward to
+diagnose. All of these exceptionts will contain the startup hook path
+(`System.StartupHookProvider.ProcessStartupHooks`) on the stack
```

Comment:
Personally I don't like a presence of a specific stack frame as a diagnostics mechanism - too easy to miss, and somewhat hard to recognize.
It also doesn't always provide enough information.
The exceptions thrown during execution of a startup hook should at the very least include which startup hook failed. Not all of the below might include this now. Also, if the hook itself throws for whatever reason, the information will only exist on the callstack.
.NET already has a similar concept in static .ctors. If there's an exception thrown from one of them we throw TypeInitializationException which wraps the actual exception inside. We could introduce similar StartupHookException as well.

Reply:
As discussed, `ProcessStartupHooks` in the stack trace isn't the only indication - most of the exception messages will include information that makes it easy to tell which hook is failing. We can still consider wrapping exceptions from the startup hook in a special exception, but I don't think that's necessary.